### PR TITLE
Stabilize some Result methods as const

### PR DESCRIPTION
**Pull request by CDirkx**
Stabilize the following methods of Result as const:
 - `is_ok`
 - `is_err`
 - `as_ref`

A test is also included, analogous to the test for `const_option`.

These methods are currently const under the unstable feature `const_result` (tracking issue: #67520).
I believe these methods to be eligible for stabilization because of the stabilization of #49146 (Allow if and match in constants) and the trivial implementations, see also: [PR#75463](https://github.com/rust-lang/rust/pull/75463) and [PR#76135](https://github.com/rust-lang/rust/pull/76135).

Note: these methods are the only methods currently under the `const_result` feature, thus this PR results in the removal of the feature.

Related: #76225

### CODE REVIEW

**Comment by CDirkx** [comment]
Created an issue for similar stabilizations: #76225

**Comment by KodrAus** [comment]
For reference, there's also an equivalent push to constify some `Result` methods at #76135

@rfcbot fcp merge

**Comment by KodrAus** [comment]
@rfcbot fcp merge

**Comment by josephlr** [comment]
There are some other methods we should also make const (either in this PR or a followup). They all have trivial match implementations and doesn't use any trait bounds:

[`Result::ok()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.ok)
[`Result::err()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.err)
[`Result::iter()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.iter)
[`Result::and()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.and)
[`Result::or()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.or)
[`Result::unwrap_or()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.unwrap_or)
[`Result::transpose()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.transpose)

[`Result::flatten()`](https://doc.rust-lang.org/nightly/core/result/enum.Result.html#method.flatten) also has a trivial impl, but that's unstable anyway (see https://github.com/rust-lang/rust/issues/70142#issuecomment-691453040)

**Review by dtolnay — APPROVED** [review]

**Comment by ecstatic-morse** [comment]
@CDirkx Test failures are legit. Clippy has a lint that suggests `is_err()` instead of `if let Err(_)`, but it's disabled in `const fn`. With this PR, that exception is no longer necessary and the test output can be updated.

**Comment by CDirkx** [comment]
I updated the testcases, when this get's merged I'll update the testcases using `Option` as well and sync these changes to the Clippy repo.

**Comment by RalfJung** [comment]
How can it be enough to update the testcases, doesn't the clippy code emitting these lints also need to be updated?

**Comment by ebroto** [comment]
Clippy is checking if the suggested methods are `const fn` in const contexts, so it works without changing the logic because they are after this change.

But once the `Option` methods are stabilized as `const fn` the function `can_suggest` and calls to it should be removed as there would be no special cases left.

**Comment by CDirkx** [comment]
That's what I thought as well (https://github.com/rust-lang/rust/pull/76135#issuecomment-695358883), but the implementation of `can_suggest` is not specific to `is_err` etc., and thus could be used in the future if Clippy were to ever suggest other unstable const methods. 

https://github.com/rust-lang/rust/blob/a3bc0e752fad96f537b73f4e9bc805a73d404f7b/src/tools/clippy/clippy_lints/src/matches.rs#L1649-L1670 

However it would be unused now, so indeed maybe better to remove it.

File: src/tools/clippy/clippy_lints/src/matches.rs [hunk 1]
```diff
@@ -1562,8 +1562,8 @@ mod redundant_pattern_match {
                             &paths::RESULT_ERR,
                             "is_ok()",
                             "is_err()",
-                            || can_suggest(cx, hir_id, sym!(result_type), "is_ok"),
-                            || can_suggest(cx, hir_id, sym!(result_type), "is_err"),
+                            || true,
+                            || true
```
Comment (RalfJung) [inline src/tools/clippy/clippy_lints/src/matches.rs:1566]:
```suggestion
                            || true,
```

**Pull request merged**